kernel/copy.masm

```diff
@@ -0,0 +1,256 @@
+  dup.1 push.0.0.0
+
+  # set initial state
+  # stack: [A, B, C, writePtr, words_even, needs_padding, ...]
+  padw padw
+
+  # stack: [A, B, C, writePtr, needs_padding, ...]
+  exec.move_from_tape_to_memory_even_words
+
+  # stack: [needs_padding, A, B, C, writePtr, ...]
+  movup.13
+
+  # if(needs_padding)
+  if.true 
+    # Rescue Prime Optimized uses overwrite mode. Drop `A`.
+    # stack: [B, C, writePtr, ...]
+    dropw
+
+    # Overwrite the `B` with the new data
+    # stack: [B', C, writePtr, ...]
+    adv_loadw
+
+    # - get the memory address that B' should be saved to
+    # - update the writePtr to point to the next address
+    # stack: [writePtr, B', writePtr+1, C, ...]
+    movup.8 dup.0 add.1 movdn.5
+
+    # save data to memory
+    # stack: [B', writePtr+1, C, ...]
+    mem_storew
+
+    # Fix writePtr position
+    # stack: [B', C, writePtr+1, ...]
+    movup.4 movdn.8
+
+    # Push padding word
+    # stack: [A, B', C, writePtr+1, ...]
+    push.1.0.0.0
+
+    # Run RPO permutation
+    # stack: [A', B', C', writePtr+1, ...]
+    rpperm
+  end
+
+  # The RPO hash is a single word, discard the unused one and the capacity.
+  # stack: [RpoHash, writePtr', ...]
+  dropw swapw dropw
+end
+
+# move_from_tape_to_memory_with_commitment(words: Felt, writePtr: Address, commitment: Commitment) -> [writePtr']
+#
+# Moves the `words` words from the advice tape to memory, starting at
+# `writePtr` address, at the end asserts the commitment data was copied
+proc.move_from_tape_to_memory_with_commitment.0
+  # Copies the tape data to memory
+  # stack: [RpoHash, writePtr', Commitment, ...]
+  exec.move_from_tape_to_memory
+
+  # Save the writePtr
+  # stack: [RpoHash, Commitment, writePtr', ...]
+  movup.4 movdn.8
+
+  # Check the commitment
+  # stack: [writePtr', ...]
+  exec.assert_eqw
+end
+
+# memcopy(readPtr: Address, writePtr: Address, words: Felt)
+#
+# Copies `words` words from `readPtr` to `writePtr`.
+proc.memcopy.0
+  # input: [readPtr, writePtr, words, ...]
+  # output: [...]
+
+  # LOOP: [readPtr, writePtr, words, ...]
+  # while(words>0)
+  dup.3 neq.0 while.true
+    # prepare read
+    # stack: [readPtr, 0, 0, 0, 0, readPtr, writePtr, words, ...]
+    padw dup.4
+
+    # perform read
+    # stack: [A, readPtr, writePtr, words, ...]
+    mem_loadw
+
+    # prepare write
+    # stack: [writePtr, A, readPtr, writePtr, words, ...]
+    dup.5
+
+    # peform write
+    # stack: [A, readPtr, writePtr, words, ...]
+    mem_storew
+
+    # clean stack
+    # stack: [readPtr, writePtr, words, ...]
+    dropw
+
+    # update counters
+    # LOOP: [readPtr+1, writePtr+1, words-1, ...]
+    inc.1 movup.2 sub.1 movup.2 inc.1 movup.2
+
+    dup.3 neq.0 # while(words>0)
+  end
+
+  # clean stack
+  # stack: [...]
+  drop drop drop
+end
+
+# move_from_stack_to_memory(writePtr: Address, words: Felt)
+#
+# Moves the `words * 4` elements from the stack to memory, starting at
+# `writePtr` address
+proc.move_from_stack_to_memory.0
```

Comment:
This is a cool procedure, but I'm not sure which context it will be useful in.

Reply:
It is to copy the global inputs from the stack to the memory. Since that section is not fully spec'ed, I opted to write a small procedure to copy the data, so that I can just update the counter if we do some changes. My idea here was to optimize for flexibility, so that we can iterate over the format quicker, instead of optimize for efficiency in number of cycles.